From a4b72a07e825e4b8c2db93a23941f31691f0bb39 Mon Sep 17 00:00:00 2001
From: Foxes Hung <foxes687@andestech.com>
Date: Tue, 28 Nov 2023 13:59:53 +0800
Subject: [PATCH 127/241] gdbstub:   Fixed empty ACR register .ace case issue.
 target/riscv:   Fixed get/set ACR value via gdb command issue since we had
 changed run ace instuction API

Signed-off-by: Foxes Hung <foxes687@andestech.com>
---
 gdbstub/andes_ace_gdb.c.inc | 10 +++++-----
 target/riscv/gdbstub.c      | 26 +++++++++++++++++++++-----
 2 files changed, 26 insertions(+), 10 deletions(-)

diff --git a/gdbstub/andes_ace_gdb.c.inc b/gdbstub/andes_ace_gdb.c.inc
index 4d8ec9ebec..87a65d40d7 100644
--- a/gdbstub/andes_ace_gdb.c.inc
+++ b/gdbstub/andes_ace_gdb.c.inc
@@ -54,19 +54,19 @@ int32_t gdb_ace_load_lib(const char *so_name)
         (unsigned *) dlsym(handle, "ace_lib_for_gdb_len");
     LE("unable to load symbol ace_lib_for_gdb_len");
 
+    /*
+     * Below ACR relative symbols can be skip since some .ace file
+     * has no any ACR
+     */
     global_acr_reg_count_v5 = (unsigned *) dlsym(handle, "acr_reg_count");
-    LE("unable to load symbol acr_reg_count");
 
     global_acr_type_count_v5 = (unsigned *) dlsym(handle, "acr_type_count");
-    LE("unable to load symbol acr_type_count");
 
     gen_get_value_code = dlsym(handle, "gen_get_value_code");
-    LE("unable to load symbol gen_get_value_code");
 
     gen_set_value_code = dlsym(handle, "gen_set_value_code");
-    LE("unable to load symbol gen_set_value_code");
 
-    if (*global_acr_type_count_v5 != 0) {
+    if (global_acr_type_count_v5 && *global_acr_type_count_v5 != 0) {
         global_acr_info_list_v5 = (ACR_INFO_T_V5 *) dlsym(handle, "acr_list");
     }
 
diff --git a/target/riscv/gdbstub.c b/target/riscv/gdbstub.c
index aee7672694..75b97b29a2 100644
--- a/target/riscv/gdbstub.c
+++ b/target/riscv/gdbstub.c
@@ -21,6 +21,8 @@
 #include "exec/gdbstub.h"
 #include "gdbstub/helpers.h"
 #include "cpu.h"
+#include "gdbstub/andes_ace_gdb.h"
+#include "andes_ace_helper.h"
 
 struct TypeSize {
     const char *gdb_type;
@@ -262,9 +264,7 @@ static int riscv_gen_dynamic_csr_xml(CPUState *cs, int base_reg)
     return CSR_TABLE_SIZE;
 }
 
-#include "gdbstub/andes_ace_gdb.h"
 unsigned char *acr_value;
-int qemu_ace_agent_run_insn(CPURISCVState *, unsigned);
 
 static int nds_ace_get_reg(CPURISCVState *env, int reg_n)
 {
@@ -276,6 +276,10 @@ static int nds_ace_get_reg(CPURISCVState *env, int reg_n)
     uint32_t curr_reg_no = 0;
     ACR_INFO_T_V5 *acr_info_list_v5 = global_acr_info_list_v5;
 
+    /* Check ACR get function whether exist */
+    if (gen_get_value_code == NULL) {
+        return -1;
+    }
     /*
      * Backup temp register (x5, x6, x7)
      * x5: high part
@@ -340,7 +344,11 @@ static int nds_ace_get_reg(CPURISCVState *env, int reg_n)
         }
 
         /* execute the code */
-        int exec_out = qemu_ace_agent_run_insn(env, insn);
+#ifdef CONFIG_USER_ONLY
+        int exec_out = qemu_ace_agent_run_insn(env, insn, 0);
+#else
+        int exec_out = qemu_ace_agent_run_insn(env, insn, env->mhartid);
+#endif
         if (exec_out != 0) {
             qemu_printf("Unable to execute ACE utility program\n");
 
@@ -386,6 +394,10 @@ static int nds_ace_set_reg(CPURISCVState *env, unsigned char *val, int reg_n)
     uint32_t curr_reg_no = 0;
     ACR_INFO_T_V5 *acr_info_list_v5 = global_acr_info_list_v5;
 
+    /* Check ACR get function whether exist */
+    if (gen_set_value_code == NULL) {
+        return -1;
+    }
     /*
      * Backup temp register (x5, x6, x7)
      * x5: high part
@@ -428,7 +440,7 @@ static int nds_ace_set_reg(CPURISCVState *env, unsigned char *val, int reg_n)
     INSN_CODE_T_V5 *insn_code = gen_set_value_code(type_name, reg_idx);
 
     bool init_acm_addr = false;
-    /* Execute the code generated by gen_get_value_code() iteratively */
+    /* Execute the code generated by gen_set_value_code() iteratively */
     for (unsigned i = 0; i < insn_code->num; i++) {
         /*
          * determine the number of GPRs used to read/write data from/to ACR/ACM
@@ -476,7 +488,11 @@ static int nds_ace_set_reg(CPURISCVState *env, unsigned char *val, int reg_n)
         unsigned insn = (insn_code->code + i)->insn;
 
         /* execute the code */
-        int exec_out = qemu_ace_agent_run_insn(env, insn);
+#ifdef CONFIG_USER_ONLY
+        int exec_out = qemu_ace_agent_run_insn(env, insn, 0);
+#else
+        int exec_out = qemu_ace_agent_run_insn(env, insn, env->mhartid);
+#endif
         if (exec_out != 0) {
             qemu_printf("Unable to execute program\n");
             /* Restore temp register */
-- 
2.34.1

